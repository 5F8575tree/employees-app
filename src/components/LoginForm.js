import React, { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useNavigate } from "react-router-dom";
import { setAuthedUser } from "../features/authedUser";

const LoginForm = () => {
  const [userName, setUserName] = useState("");
  const [password, setPassword] = useState("");
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const userList = useSelector((state) => state.users.users);

  const handleSubmit = (e) => {
    e.preventDefault();

    const correct_login = () => {
      console.log("login successfull")
      dispatch(setAuthedUser(userName));

      localStorage.setItem("authedUser", userName);
      navigate("/Dashboard");
    }

    const usernames = Object.values(userList)
    usernames.map(v => v.id === userName ? v.password === password ? correct_login() : console.log("wrong password") : null)

    // If input matches a user in the userList and password matches the password in the userList, set the authedUser to the user's id
    // if (userList.userName.password === password) {
    //   dispatch(setAuthedUser(userName));

    //   localStorage.setItem("authedUser", userName);
    //   navigate("/Dashboard");
    // }
    // else { console.log("wrong credentials") }
  };

  // console.log(userList.mtsamis.password)

  return (
    <div className="split_screen">
      { }
      <div className="right">
        <section className="copy">
          <div className="center">
            <div className="login-form">
              <h1 data-testid="heading">Login</h1>
              <form method="post" onSubmit={handleSubmit}>
                <div className="txt_field">
                  <input
                    id="usernameField"
                    type="text"
                    required
                    value={userName}
                    onChange={(e) => setUserName(e.target.value)}
                  ></input>
                  <span></span>
                  <label htmlFor="usernameField">Username</label>
                </div>
                <div className="txt_field">
                  <input
                    id="passwordField"
                    type="password"
                    required
                    value={password}
                    onChange={(e) => setPassword(e.target.value)}
                  ></input>
                  <span></span>
                  <label htmlFor="passwordField">Password</label>
                </div>
                <div className="pass">Forgot Password?</div>
                <input type="submit" value="Login" data-testid="submitButton"></input>
                <div className="signup_link" data-testid="signupLink">
                  Not a member? <a href="/">Signup</a>
                </div>
                <div className="copy_legal">
                  <p>
                    <span className="small" data-testid="copyrightNotice">
                      Copyright 2022 | View our <a href="/">Privacy Policy</a>{" "}
                      &amp; <a href="/">Terms and Conditions</a>
                    </span>
                  </p>
                </div>
              </form>
            </div>
          </div>
        </section>
      </div>
    </div>
  );
};

export default LoginForm;
