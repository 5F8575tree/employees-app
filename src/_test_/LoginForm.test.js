import { React } from 'react';
import { screen, render, fireEvent } from '@testing-library/react';

import LoginForm from '../components/LoginForm';

import '@testing-library/jest-dom';

const localStorageMock = (function() {
  let store = {}

  return {
    getItem: function(key) {
      return store[key] || null
    },
    setItem: function(key, value) {
      store[key] = value.toString()
    },
    removeItem: function(key) {
      delete store[key]
    },
    clear: function() {
      store = {}
    }
  }
})()

Object.defineProperty(window, 'localStorage', {
  value: localStorageMock
})


const mockData = {
  dispatch: jest.fn(),
  navigate: jest.fn(),
  state: {
    users: {
      users: {
        0: {
          id: 'test-user',
          password: 'secret123'
        },
      }
    }
  }
}

jest.mock("react-redux", () => ({
  useDispatch: () => mockData.dispatch,
  useSelector: (cb => {
    const result = cb(mockData.state)
    return result;
  })
}))

jest.mock("react-router-dom", () => ({
  useNavigate: () => mockData.navigate
}))

describe('LoginForm', () => {
  /**
   * Snapshot Testing
   */
  it("renders the LoginForm component snapshot", () => {
    const { asFragment } = render(
      <LoginForm />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  /**
   * Check the form is displayed with correct fields
   */
  it('should display login form to the user',  () => {
    render(<LoginForm />)

    // heading or title
    const heading = screen.getByTestId('heading');
    expect(heading).toBeInTheDocument()

    // username label and field
    const usernameLabel = screen.getByText('Username');
    expect(usernameLabel).toBeInTheDocument();

    const usernameField = screen.getByLabelText('Username');
    expect(usernameField).toBeInTheDocument();

    // password label and field
    const passwordLabel = screen.getByText('Password');
    expect(passwordLabel).toBeInTheDocument();

    const passwordField = screen.getByLabelText('Password');
    expect(passwordField).toBeInTheDocument();

    // submit button
    const submitButton = screen.getByTestId('submitButton');
    expect(submitButton).toBeInTheDocument();

    // signup link
    const signupLink = screen.getByTestId('signupLink');
    expect(signupLink).toBeInTheDocument();

    // copyright notice
    const copyrightNotice = screen.getByTestId('copyrightNotice');
    expect(copyrightNotice).toBeInTheDocument();
  })

  /**
   * Enter username and password in the form
   */
  it('allows user to enter data in input fields', () => {
    render(<LoginForm />)

    const mockedCreds = {
      username: 'test-user',
      password: 'test-123'
    }

    const usernameField = screen.getByLabelText('Username');
    expect(usernameField).toBeInTheDocument();

    fireEvent.input(usernameField, {
      target: {
        value: mockedCreds.username
      }
    })

    const passwordField = screen.getByLabelText('Password');
    expect(passwordField).toBeInTheDocument();

    fireEvent.input(passwordField, {
      target: {
        value: mockedCreds.password
      }
    })

    const filledUsernameField = screen.getByDisplayValue(mockedCreds.username);
    expect(filledUsernameField).toBeInTheDocument();

    const filledPasswordField = screen.getByDisplayValue(mockedCreds.password);
    expect(filledPasswordField).toBeInTheDocument();
  })

  /**
   * Validate the form before submitting
   */
  it('should validate the user inputs before submitting the form', () => {
    render(<LoginForm />)

    const usernameField = screen.getByLabelText('Username');
    expect(usernameField).toBeInTheDocument();

    const passwordField = screen.getByLabelText('Password');
    expect(passwordField).toBeInTheDocument();

    const submitButton = screen.getByTestId('submitButton');
    expect(submitButton).toBeInTheDocument();

    fireEvent.click(submitButton);

    expect(mockData.dispatch).not.toHaveBeenCalled()
  });

  /**
   * Validate user credentials before redirecting
   */
   it('should validate the credentials before redirecting to dashboard ', () => {
    render(<LoginForm />)

    const mockedCreds = {
      username: 'test-user',
      password: 'secret1234'
    }

    const consoleLogSpy = jest.spyOn(console, 'log');

    const usernameField = screen.getByLabelText('Username');
    expect(usernameField).toBeInTheDocument();

    fireEvent.input(usernameField, {
      target: {
        value: mockedCreds.username
      }
    })

    const passwordField = screen.getByLabelText('Password');
    expect(passwordField).toBeInTheDocument();

    fireEvent.input(passwordField, {
      target: {
        value: mockedCreds.password
      }
    })

    const submitButton = screen.getByTestId('submitButton');
    expect(submitButton).toBeInTheDocument();

    fireEvent.click(submitButton);

    expect(consoleLogSpy).toHaveBeenCalledWith("wrong password");
  });

  /**
   * Show allow user to submit the form after entering the valid values
   */
   it('should allow the user to submit the form with valid credentials', () => {
    const mockedCreds = {
      username: 'test-user',
      password: 'secret123'
    }

    const localStorageSpy = jest.spyOn(window.localStorage, 'setItem');

    render(<LoginForm />)

    const usernameField = screen.getByLabelText('Username');
    expect(usernameField).toBeInTheDocument();

    fireEvent.input(usernameField, {
      target: {
        value: mockedCreds.username
      }
    })

    const passwordField = screen.getByLabelText('Password');
    expect(passwordField).toBeInTheDocument();

    fireEvent.input(passwordField, {
      target: {
        value: mockedCreds.password
      }
    })

    const submitButton = screen.getByTestId('submitButton');
    expect(submitButton).toBeInTheDocument();

    fireEvent.click(submitButton);

    expect(mockData.dispatch).toHaveBeenCalled();
    expect(mockData.navigate).toHaveBeenCalledWith("/Dashboard");
    expect(localStorageSpy).toHaveBeenCalledWith("authedUser", mockedCreds.username)
  });
})