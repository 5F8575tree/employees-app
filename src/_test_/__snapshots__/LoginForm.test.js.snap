// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginForm renders the LoginForm component snapshot 1`] = `
<DocumentFragment>
  <div
    class="split_screen"
  >
    <div
      class="right"
    >
      <section
        class="copy"
      >
        <div
          class="center"
        >
          <div
            class="login-form"
          >
            <h1
              data-testid="heading"
            >
              Login
            </h1>
            <form
              method="post"
            >
              <div
                class="txt_field"
              >
                <input
                  id="usernameField"
                  required=""
                  type="text"
                  value=""
                />
                <span />
                <label
                  for="usernameField"
                >
                  Username
                </label>
              </div>
              <div
                class="txt_field"
              >
                <input
                  id="passwordField"
                  required=""
                  type="password"
                  value=""
                />
                <span />
                <label
                  for="passwordField"
                >
                  Password
                </label>
              </div>
              <div
                class="pass"
              >
                Forgot Password?
              </div>
              <input
                data-testid="submitButton"
                type="submit"
                value="Login"
              />
              <div
                class="signup_link"
                data-testid="signupLink"
              >
                Not a member? 
                <a
                  href="/"
                >
                  Signup
                </a>
              </div>
              <div
                class="copy_legal"
              >
                <p>
                  <span
                    class="small"
                    data-testid="copyrightNotice"
                  >
                    Copyright 2022 | View our 
                    <a
                      href="/"
                    >
                      Privacy Policy
                    </a>
                     & 
                    <a
                      href="/"
                    >
                      Terms and Conditions
                    </a>
                  </span>
                </p>
              </div>
            </form>
          </div>
        </div>
      </section>
    </div>
  </div>
</DocumentFragment>
`;
